### PR TITLE
Refactoring to prepare for the addition of dynamic fast field

### PR DESCRIPTION
**Pull request by fulmicoton**
- Exposing insert_key / insert_value
- Renamed SSTable::{Reader/Writer}-> SSTable::{ValueReader/ValueWriter}
- Added a generic Dictionary object in the sstable crate
- Removing the TermDictionary wrapper from tantivy, relying directly on an alias of the generic Dictionary object.
- dropped the use of byteorder in sstable.
- Stopped scanning / reading the entire dictionary when streaming a range.

### CODE REVIEW

File: sstable/src/dictionary.rs [hunk 1]
```diff
@@ -1,150 +1,121 @@
...
-        .expect("Writing in a Vec<u8> should never fail");
-    FileSlice::from(term_dictionary_data)
-});
+use crate::streamer::{Streamer, StreamerBuilder};
+use crate::{BlockAddr, DeltaReader, Reader, SSTable, SSTableIndex, TermOrdinal};
 
 /// The term dictionary contains all of the terms in
 /// `tantivy index` in a sorted manner.
 ///
 /// The `Fst` crate is used to associate terms to their
```
Comment (PSeitz) [inline sstable/src/dictionary.rs]: I think that fst comment is outdated, would be nice to explain what the `SSTable` is used for

File: sstable/src/dictionary.rs [hunk 2]
```diff
@@ -1,150 +1,121 @@
...
+        Ok(TSSTable::delta_reader(data))
+    }
+
+    fn file_slice_for_range(&self, key_range: impl RangeBounds<[u8]>) -> FileSlice {
+        let start_bound: Bound<usize> = match key_range.start_bound() {
+            Bound::Included(key) | Bound::Excluded(key) => {
+                let Some(first_block_addr) = self.sstable_index.search_block(key) else {
+                    return FileSlice::empty();
+                };
+                Bound::Included(first_block_addr.byte_range.start)
```
Comment (PSeitz) [inline sstable/src/dictionary.rs:101]: a comment why `Bound::Excluded` and `Bound::Included` map to `Bound::Included(first_block_addr.byte_range.start)` would be nice

File: sstable/src/value/mod.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    /// Returns the number of bytes that were written.
+    fn load(&mut self, data: &[u8]) -> io::Result<usize>;
+}
+
+pub trait ValueWriter: Default {
+    /// Type of the value being written.
+    type Value;
+
+    /// Records a new value.
+    /// This method usually just accumulates data in a `Vec`,
```
Comment (PSeitz) [inline sstable/src/value/mod.rs:30]: great comment

File: sstable/src/value/mod.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    /// only to be serialized on the call to `ValueWriter::write_block`.
```
Comment (PSeitz) [inline sstable/src/value/mod.rs]:
```suggestion
    /// only to be serialized on the call to `ValueWriter::serialize_block`.
```

File: sstable/src/value/mod.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    /// Serializes the accumulated values into the output buffer.
+    fn serialize_block(&mut self, output: &mut Vec<u8>);
+}
+
+pub use range::{RangeReader, RangeWriter};
+pub use u64_monotonic::{U64MonotonicReader, U64MonotonicWriter};
+pub use void::{VoidReader, VoidWriter};
+
+fn deserialize_u64(data: &mut &[u8]) -> u64 {
```
Comment (PSeitz) [inline sstable/src/value/mod.rs]:
```suggestion
fn deserialize_vint_u64(data: &mut &[u8]) -> u64 {
```

File: sstable/src/value/range.rs [hunk 6]
```diff
@@ -0,0 +1,95 @@
+use std::io;
+use std::ops::Range;
+
+use crate::value::{deserialize_u64, ValueReader, ValueWriter};
+
+#[derive(Default)]
+pub struct RangeReader {
```
Comment (PSeitz) [inline sstable/src/value/range.rs]: a comment to describe what this does and what it's used for would be good

Comment (PSeitz) [inline sstable/src/value/range.rs]: I would use slightly more verbose names
RangeValueReader, VoidValueReader, U64MonotonicValueReader

**Review by PSeitz — APPROVED** [review]
a lot of rename improvements, nice

I left some minor comments

File: sstable/src/dictionary.rs (hunk 1 above)
Comment (fulmicoton) [reply sstable/src/dictionary.rs]: thanks!

File: sstable/src/value/mod.rs (hunk 5 above)
Comment (fulmicoton) [reply sstable/src/value/mod.rs]: renamed

File: sstable/src/value/range.rs (hunk 6 above)
Comment (fulmicoton) [reply sstable/src/value/range.rs]: added comment

**Pull request merged**